bindings/ergo-lib-wasm/src/merkleproof.rs

```diff
@@ -0,0 +1,93 @@
+//! Merkle Proof verification
+use serde::{Deserialize, Serialize};
+use std::convert::TryInto;
+use wasm_bindgen::prelude::*;
+
+#[wasm_bindgen]
+/// A MerkleProof type. Given leaf data and levels (bottom-upwards), the root hash can be computed and validated
+pub struct MerkleProof(ergo_merkle_tree::MerkleProof);
+
+#[wasm_bindgen]
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, Serialize, Deserialize)]
+/// The side the merkle node is on in the tree
+pub enum NodeSide {
+    /// Node is on the left side of current level
+    Left = 0u8,
+    /// Node is on the right side of current level
+    Right = 1u8,
+}
+
+impl From<NodeSide> for ergo_merkle_tree::NodeSide {
+    fn from(side: NodeSide) -> ergo_merkle_tree::NodeSide {
+        match side {
+            NodeSide::Left => ergo_merkle_tree::NodeSide::Left,
+            NodeSide::Right => ergo_merkle_tree::NodeSide::Right,
+        }
+    }
+}
+impl From<ergo_merkle_tree::NodeSide> for NodeSide {
+    fn from(side: ergo_merkle_tree::NodeSide) -> NodeSide {
+        match side {
+            ergo_merkle_tree::NodeSide::Left => NodeSide::Left,
+            ergo_merkle_tree::NodeSide::Right => NodeSide::Right,
+        }
+    }
+}
+
+/// A level node in a merkle proof
+#[wasm_bindgen]
+#[derive(Serialize, Deserialize, Debug, Copy, Clone)]
+pub struct LevelNode(ergo_merkle_tree::LevelNode);
+
+#[wasm_bindgen]
+impl LevelNode {
+    /// Creates a new LevelNode from a 32 byte hash and side that the node belongs on in the tree. Fails if the digest is not 32 bytes
+    pub fn new(hash: &[u8], side: NodeSide) -> Result<LevelNode, JsValue> {
```

Comment:
Passed `side` becomes null in JS after this call. It's better to pass by reference. The same goes to `MerkleProof::add_node` below.

Reply:
wasm_bindgen doesn't seem to have an implementation of RefFromWasmAbi for #[repr(u8)] enums. As an alternative, I've decided to pass a u8 instead, since NodeSide implements TryFrom<u8>. I'm not too familiar with wasm_bindgen, so let me know if there is a better way!